app/src/test/java/com/futurice/freesound/network/api/DefaultFreeSoundApiServiceTest.kt

```diff
@@ -0,0 +1,198 @@
+
+        defaultFreeSoundApiService = DefaultFreeSoundApiService(freeSoundApi,
+                TEST_CLIENT_ID,
+                TEST_CLIENT_SECRET)
+    }
+
+    @Test
+    fun getUser_emitsUser_whenApiSuccessful() {
+        arrange {
+            user { DUMMY_USER }
+        }
+
+        defaultFreeSoundApiService.getUser("username")
+                .test()
+                .assertValue(DUMMY_USER)
+    }
+
+    @Test
+    fun getUser_invokesApiWithUsernameParameter() {
+        arrange {
+            user { DUMMY_USER }
+        }
+        val username = "username"
+
+        defaultFreeSoundApiService.getUser(username).subscribe()
+
+
+        verify<FreeSoundApi>(freeSoundApi).user(username)
+    }
+
+    @Test
+    fun getUser_emitsError_whenApiError() {
+        arrange {
+            userError { ERROR }
+        }
+
+        defaultFreeSoundApiService.getUser("username")
+                .test()
+                .assertError(ERROR)
+    }
+
+    @Test
+    fun getAccessToken_emitsAccessToken_whenApiSuccessful() {
+        arrange {
+            token { DUMMY_ACCESS_TOKEN }
+        }
+        defaultFreeSoundApiService.getAccessToken("code")
+                .test()
+                .assertValue(DUMMY_ACCESS_TOKEN)
+    }
+
+    @Test
+    fun getAccessToken_invokesApiWithCorrectParameters() {
+        arrange {
+            token { DUMMY_ACCESS_TOKEN }
+        }
+        val code = "code"
+
+        defaultFreeSoundApiService.getAccessToken(code).subscribe()
+
+
+        verify<FreeSoundApi>(freeSoundApi).accessToken(TEST_CLIENT_ID,
+                TEST_CLIENT_SECRET,
+                ApiConstants.AUTHORIZATION_CODE_GRANT_TYPE_VALUE,
+                code)
+    }
+
+    @Test
+    fun getAccessToken_emitsError_whenApiError() {
+        arrange {
+            tokenError { ERROR }
+        }
+
+        defaultFreeSoundApiService.getAccessToken("code")
+                .test()
+                .assertError(ERROR)
+    }
+
+    @Test
+    fun search_emitsResults_whenApiSuccessful() {
+        arrange {
+            search { TEST_SEARCH_RESULT }
+        }
+
+        defaultFreeSoundApiService.search("query")
+                .test()
+                .assertValue(TEST_SEARCH_RESULT)
+    }
+
+    @Test
+    fun search_invokesApiWithCorrectParameters() {
+        arrange {
+            search { TEST_SEARCH_RESULT }
+        }
+        val query = "query"
+        defaultFreeSoundApiService.search(query).subscribe()
+
+        verify<FreeSoundApi>(freeSoundApi).search(eq(query), isNull<String>(), eq(SoundFields.BASE))
+    }
+
+    @Test
+    fun search_emitsError_whenApiError() {
+        arrange {
+            searchError { ERROR }
+        }
+
+        defaultFreeSoundApiService.search("query")
+                .test()
+                .assertError(ERROR)
+    }
+
+    fun arrange(init: Arrangement.() -> Unit) = Arrangement().apply { init() }
+
+    inner class Arrangement {
+
+        fun search(init: () -> SoundSearchResult): Any =
```

Comment:
I wanted to avoid having a `{ return... }` function and preferred `=` function.
But `when` returns something that is not a `Unit`.
I am not yet sold on my solution, so if you do not like it, I can change it